### PR TITLE
implemented normalization and leiden algorithm 

### PR DESCRIPTION
**Pull request by jchpan**

### CODE REVIEW

File: appyters/dimensionality_reduction_visualization/dimensionality_reduction_visualization.ipynb [hunk 1]
```diff
@@ -238,7 +272,46 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "expr_df.head()"
+    "# check if data have been loaded correctly\n",
+    "try:\n",
+    "    pca = PCA(n_components=clustering_topk)\n",
+    "    pca.fit_transform(expr_df)\n",
+    "except:\n",
+    "    raise Exception(\"Error! Incorrect format. Please check if the data have been read/seperated correctly.\")"
```
Comment (u8sand) [inline appyters/dimensionality_reduction_visualization/dimensionality_reduction_visualization.ipynb:280]: Any reason you do this twice if normalizing data?

I think you can just take the PCA out of the if needs_normalize below

File: appyters/dimensionality_reduction_visualization/dimensionality_reduction_visualization.ipynb [hunk 2]
```diff
@@ -351,9 +440,23 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "%%appyter code_exec\n",
     "counter = 1\n",
     "\n",
+    "if need_leiden:\n",
+    "    leiden_visualization()\n",
```
Comment (u8sand) [inline appyters/dimensionality_reduction_visualization/dimensionality_reduction_visualization.ipynb:446]: Just put the contents of this function here instead of using it or make the function pure (i.e. all variables it depends on in the arguments of the function)

File: appyters/dimensionality_reduction_visualization/dimensionality_reduction_visualization.ipynb [hunk 3]
```diff
@@ -387,12 +490,12 @@
     "    caption = \"t-SNE 2D scatter plot. Each point presents a column from the input data matrix.\"\n",
     "\n",
     "diplay_caption(counter, caption)\n",
+    "counter += 1\n",
```
Comment (u8sand) [inline appyters/dimensionality_reduction_visualization/dimensionality_reduction_visualization.ipynb:493]: A fun way to make this slightly more pythonic is to use

```python
# at the top
import itertools
counter = iter(itertools.count(start=1))


# everywhere else
display_caption(next(counter), caption)
```

File: appyters/dimensionality_reduction_visualization/dimensionality_reduction_visualization.ipynb [hunk 4]
```diff
@@ -229,7 +261,9 @@
     "if has_matadata:\n",
```
Comment (u8sand) [inline appyters/dimensionality_reduction_visualization/dimensionality_reduction_visualization.ipynb:261]: This is old but looks misspelled?

**Review by u8sand — CHANGES_REQUESTED** [review]

File: appyters/dimensionality_reduction_visualization/dimensionality_reduction_visualization.ipynb (hunk 3 above)
Comment (jchpan) [reply appyters/dimensionality_reduction_visualization/dimensionality_reduction_visualization.ipynb]: Great idea! Just changed it.

File: appyters/dimensionality_reduction_visualization/dimensionality_reduction_visualization.ipynb (hunk 1 above)
Comment (jchpan) [reply appyters/dimensionality_reduction_visualization/dimensionality_reduction_visualization.ipynb:282]: Sometimes when input data has been normalized, we get many NA values after normalization, so I was checking if the input data has already been normalized.

File: appyters/dimensionality_reduction_visualization/dimensionality_reduction_visualization.ipynb [hunk 5]
```diff
@@ -254,15 +328,18 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "def diplay_caption(counter, caption):\n",
-    "    display(Markdown(\"*Figure {}. {}*\".format(counter, caption)))\n",
+    "counter = iter(itertools.count(start=0))\n",
```
Comment (u8sand) [inline appyters/dimensionality_reduction_visualization/dimensionality_reduction_visualization.ipynb:331]: The first figure will be Figure 0 if you start at 0

File: appyters/dimensionality_reduction_visualization/dimensionality_reduction_visualization.ipynb [hunk 6]
```diff
@@ -345,14 +441,23 @@
...
    ]
   },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "caption = \"Leiden 2D scatter plot. Each point presents a column from the input data matrix.\"\n",
+    "rna = leiden_visualization(need_leiden, caption)"
```
Comment (u8sand) [inline appyters/dimensionality_reduction_visualization/dimensionality_reduction_visualization.ipynb:451]: Please eliminate the `need_leiden` argument in this function and put the if statement outside..

File: appyters/dimensionality_reduction_visualization/requirements.txt [hunk 7]
```diff
@@ -7,8 +7,13 @@ pandas
...
+scanpy
 IPython
 bokeh
 umap-learn
-MulticoreTSNE git+https://github.com/maayanlab/Multicore-TSNE
-react-scatter-board @ git+https://github.com/maayanlab/react-scatter-board@v2.1
\ No newline at end of file
+scipy
+qnorm
+multiprocess
```
Comment (u8sand) [inline appyters/dimensionality_reduction_visualization/requirements.txt:16]: I don't see where this is used?

**Review by u8sand — CHANGES_REQUESTED** [review]

File: appyters/dimensionality_reduction_visualization/requirements.txt (hunk 7 above)
Comment (jchpan) [reply appyters/dimensionality_reduction_visualization/requirements.txt]: Removed multiprocessing. qnorm was for normalization and scipy was for leiden algorithm.

**Review by u8sand — APPROVED** [review]
